pkg/volume/csi/csi_mounter.go

```diff
@@ -0,0 +1,204 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"errors"
+	"fmt"
+	"path"
+
+	"github.com/golang/glog"
+	grpctx "golang.org/x/net/context"
+	api "k8s.io/api/core/v1"
+	storage "k8s.io/api/storage/v1alpha1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	kstrings "k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type csiMountMgr struct {
+	k8s        kubernetes.Interface
+	csiClient  csiClient
+	plugin     *csiPlugin
+	driverName string
+	volumeID   string
+	readOnly   bool
+	spec       *volume.Spec
+	pod        *api.Pod
+	podUID     types.UID
+	options    volume.VolumeOptions
+	volumeInfo map[string]string
+	volume.MetricsNil
+}
+
+// volume.Volume methods
+var _ volume.Volume = &csiMountMgr{}
+
+func (c *csiMountMgr) GetPath() string {
+	return getTargetPath(c.podUID, c.driverName, c.volumeID, c.plugin.host)
+}
+
+func getTargetPath(uid types.UID, driverName string, volID string, host volume.VolumeHost) string {
+	// driverName validated at Mounter creation
+	// sanitize (replace / with ~) in volumeID before it's appended to path:w
+	driverPath := fmt.Sprintf("%s/%s", driverName, kstrings.EscapeQualifiedNameForDisk(volID))
+	return host.GetPodVolumeDir(uid, kstrings.EscapeQualifiedNameForDisk(csiPluginName), driverPath)
+}
+
+// volume.Mounter methods
+var _ volume.Mounter = &csiMountMgr{}
+
+func (c *csiMountMgr) CanMount() error {
+	//TODO (vladimirvivien) use this method to probe controller using CSI.NodeProbe() call
+	// to ensure Node service is ready in the CSI plugin
+	return nil
+}
+
+func (c *csiMountMgr) SetUp(fsGroup *int64) error {
+	return c.SetUpAt(c.GetPath(), fsGroup)
+}
+
+func (c *csiMountMgr) SetUpAt(dir string, fsGroup *int64) error {
+	glog.V(4).Infof(log("Mounter.SetUpAt(%s)", dir))
+
+	ctx, cancel := grpctx.WithTimeout(grpctx.Background(), csiTimeout)
+	defer cancel()
+
+	csi := c.csiClient
+	pvName := c.spec.PersistentVolume.GetName()
+
+	// ensure version is supported
+	if err := csi.AssertSupportedVersion(ctx, csiVersion); err != nil {
+		glog.Errorf(log("failed to assert version: %v", err))
+		return err
+	}
+
+	// TODO (vladimirvivien) make assertion calls private inside client.
+	// assert volume attachment capability from driver,
+	// if not supported, assume vol is not previously attached and not ready for mounting
+	if err := csi.AssertVolumePublishCapability(ctx); err != nil {
+		glog.Error(log("failed to assert volume publish capability: %v", err))
+		return err
+	}
+
+	// search for attachment by VolumeAttachment.Spec.Source.PersistentVolumeName
+	if c.volumeInfo == nil {
+		attachList, err := c.k8s.StorageV1alpha1().VolumeAttachments().List(meta.ListOptions{})
```

Comment:
1. Since the `VolumeAttachment` name is now the `volumeName` not the `pvName` you should just be able to generate the name from the spec and Get the object directly. Right?
2. If you implement the `WaitForAttach(...)` method, as suggested above, you could just have that method cache the object on success in `csiMountMgr` for this method to use.

Reply:
1) Yes, you are correct.  2) As I understand it, these two objects, mounter and attacher, get created at different times, and called at different time by the API, so not sure that would work.